### DISCUSSION TITLE
Still under active development? And how to get started?

### DISCUSSION THREAD

**Discussion by harmolipi** [alexstandiford/underpin#59] [Q&A] [answered]
I just came across Underpin, and it seems like a dream come true for Wordpress plugin development. I've been bashing my head over how unintuitive everything is in WP, and I love how smooth and object-oriented Underpin seems to make everything.

But it looks like with the upgrade to v3.0, the main readme was gutted, and there's not much info regarding how to actually set up a new plugin with underpin. So far I've been able to figure out that I needed to maybe drop the plugin boilerplate directly into my mu-plugins directory ([https://github.com/Underpin-WP/underpin](https://github.com/Underpin-WP/underpin))? But also that I could maybe just include it in a single plugin rather than as a sitewide must-use plugin ([https://github.com/Underpin-WP/underpin/discussions/42](https://github.com/Underpin-WP/underpin/discussions/42))?

I tried both, and I'm just trying to test it out by setting up a new custom post type [with the appropriate loader](https://github.com/Underpin-WP/custom-post-type-loader), but nothing seems to happen.

```php
<?php
/*
Plugin Name: Plugin Name Replace Me
Description: Plugin Description Replace Me
Version: 1.0.0
Author: An awesome developer
Text Domain: plugin_name_replace_me
Domain Path: /languages
Requires at least: 5.1
Requires PHP: 7.0
Author URI: https://www.designframesolutions.com
*/

use Underpin\Abstracts\Underpin;

if (!defined('ABSPATH')) {
	exit;
}

// Load Underpin, and its dependencies.
$autoload = plugin_dir_path(__FILE__) . 'vendor/autoload.php';

require_once($autoload);

/**
 * Fetches the instance of the plugin.
 * This function makes it possible to access everything else in this plugin.
 * It will automatically initiate the plugin, if necessary.
 * It also handles autoloading for any class in the plugin.
 *
 * @since 1.0.0
 *
 * @return \Underpin\Factories\Underpin_Instance The bootstrap for this plugin.
 */
function plugin_name_replace_me()
{
	return Underpin::make_class([
		'root_namespace'      => 'Plugin_Name_Replace_Me',
		'text_domain'         => 'plugin_name_replace_me',
		'minimum_php_version' => '7.0',
		'minimum_wp_version'  => '5.1',
		'version'             => '1.0.0',
	])->get(__FILE__);
}

// Lock and load.
plugin_name_replace_me();

plugin_name_replace_me()->custom_post_types()->add('test_post_type', [
	'type' => 'test_post_type', // see register_post_type
	'args' => [
		'name'                  => _x('Test Post Types', 'Post Type General Name', 'text_domain'),
		'singular_name'         => _x('Test Post Type', 'Post Type Singular Name', 'text_domain'),
		'menu_name'             => __('Test Post Types', 'text_domain'),
		'name_admin_bar'        => __('Test Post Type', 'text_domain'),
	], // see register_post_type
]);
```

What am I doing wrong here? And thank you so much for this!

**Comment by alexstandiford** [comment, marked as answer]
Hey @harmolipi - Sorry, you've kind-of caught me in the midst of a transition, here.

Yep, this is still very much in active development (I just [issued a new release](https://github.com/Underpin-WP/underpin/releases/tag/4.1.0) a few days ago). I'm in the midst of refactoring this to be less-opinionated, while still remaining super useful.

The short-term simplest thing to-do in the mid-term is to downgrade to use [Underpin 2.1.0](https://github.com/Underpin-WP/underpin/releases/tag/2.1.0). It all works quite well, I just realized that I don't always want to write plugins using this pattern and wanted to expand the functionality to be able to work on different platforms, as well as with different paradigms. It works well if that's the pattern you want to use, and WordPress code doesn't change often enough that there's any concerns that it will become incompatible in the future. [This tutorial](https://wpdev.academy/wordpress-tutorials/how-to-set-up-underpin-for-a-single-site/) can show you how to set that up.

The newer version has loads of nice things, and I have been using it on personal projects using [this boilerplate](https://github.com/Underpin-WP/boilerplate-wordpress), but I haven't drawn a line in the sand to put it all together and make it cohesive yet. I just haven't found the time to do it.

I gutted the readme because none of it was true at the moment (although, I really _should_ add a note in the readme to downgrade to 2.1.0 if you're wanting to use the older approach).

**Reply by harmolipi** [reply]
Thank you for the quick reply, that all makes a lot of sense! I appreciate that you're still working on it and taking it in that direction. A note about downgrading would be handy, as well as perhaps a quick tutorial that takes you through not only setting it up like in the one you linked, but also just adding a custom post type and a custom taxonomy or something, just so people get a sense of how that works and what a viable project structure might look like. That might be helpful on both the older version and on the newer version, so people can get up and running with either pretty quickly.

In the meantime, I did manage to figure out how to add a custom post type (I think adding that particular loader forced composer to downgrade Underpin anyway), and I did really appreciate having this level of abstraction over the way WP normally has you do things. Since I'm brand new it may not be the most accurate, but if you want as I figure things out, I can submit a PR to add a simple tutorial like that to the readme, since I've got the perspective of fresh eyes on this project. But only if you think that'd be useful.

Thanks again for working on this. Every time I have to re-google how to do something super basic and common in WP, I smash my face into the wall and wonder why WP hasn't cleaned it up (i.e. made it more like Underpin) yet.

**Reply by alexstandiford** [reply]
> I can submit a PR to add a simple tutorial like that to the readme, since I've got the perspective of fresh eyes on this project. But only if you think that'd be useful.

I would _really_ appreciate that. This perspective is invaluable.

> Thanks again for working on this. Every time I have to re-google how to do something super basic and common in WP, I smash my face into the wall and wonder why WP hasn't cleaned it up (i.e. made it more like Underpin) yet.

Heck yeah! I'm glad it's helping!

**Reply by harmolipi** [reply]
Sweet, as soon as I get a chance I'll try and do that then!

**Reply by alexstandiford** [reply]
Related:

https://github.com/orgs/Underpin-WP/discussions/61
